File: LowLevelDesign/tic-tac-toe/src/observer/GameSubject.ts
import { Game } from "../Game";
import { GameObserver } from "./GameObserver";

export class GameSubject {
    private observers: GameObserver[] = [];

    public registerObserver(observer: GameObserver): void {
        this.observers.push(observer);
    }

    public unregisterObserver(observer: GameObserver): void {
        this.observers = this.observers.filter(obs => obs !== observer);
    }

    public notifyObservers(): void {
        for (const observer of this.observers) {
            // Todo: Revisit this
            observer.update(this as unknown as Game)
        }
    }
}

File: LowLevelDesign/java/tictactoe/src/com/tictac/observer/GameSubject.java
package com.tictac.observer;

import com.tictac.entities.Game;

import java.util.ArrayList;
import java.util.List;

public class GameSubject {
    private List<GameObserver> observers = new ArrayList<>();
    public void addObserver(GameObserver observer) {
        this.observers.add(observer);
    }

    public void removeObserver(GameObserver observer) {
        this.observers.remove(observer);
    }

    public void notifyObservers(Game game) {
        System.out.println(this.observers);
        for(GameObserver observer: this.observers) {
            observer.update(game);
        }
    }
}
